lib/src/request.dart

```diff
@@ -3,159 +3,94 @@
-    }
-  }
-
-  /// The form-encoded fields in the body of the request as a map from field
-  /// names to values. The form-encoded body is converted to and from
-  /// [bodyBytes] using [encoding] (in the same way as [body]).
-  ///
-  /// If the request doesn't have a `Content-Type` header of
-  /// `application/x-www-form-urlencoded`, reading this will throw a
-  /// [StateError].
-  ///
-  /// If the request has a `Content-Type` header with a type other than
-  /// `application/x-www-form-urlencoded`, setting this will throw a
-  /// [StateError]. Otherwise, the content type will be set to
-  /// `application/x-www-form-urlencoded`.
-  ///
-  /// This map should only be set, not modified in place.
-  Map<String, String> get bodyFields {
-    var contentType = _contentType;
-    if (contentType == null ||
-        contentType.mimeType != "application/x-www-form-urlencoded") {
-      throw new StateError('Cannot access the body fields of a Request without '
-          'content-type "application/x-www-form-urlencoded".');
-    }
-
-    return Uri.splitQueryString(body, encoding: encoding);
-  }
-
-  set bodyFields(Map<String, String> fields) {
-    var contentType = _contentType;
-    if (contentType == null) {
-      _contentType = new MediaType("application", "x-www-form-urlencoded");
-    } else if (contentType.mimeType != "application/x-www-form-urlencoded") {
-      throw new StateError('Cannot set the body fields of a Request with '
-          'content-type "${contentType.mimeType}".');
-    }
-
-    this.body = mapToQuery(fields, encoding: encoding);
-  }
-
-  /// Creates a new HTTP request.
-  Request(String method, Uri url)
-    : _defaultEncoding = UTF8,
-      _bodyBytes = new Uint8List(0),
-      super(method, url);
-
-  /// Freezes all mutable fields and returns a single-subscription [ByteStream]
-  /// containing the request body.
-  ByteStream finalize() {
-    super.finalize();
-    return new ByteStream.fromBytes(bodyBytes);
-  }
-
-  /// The `Content-Type` header of the request (if it exists) as a
-  /// [MediaType].
-  MediaType get _contentType {
-    var contentType = headers['content-type'];
-    if (contentType == null) return null;
-    return new MediaType.parse(contentType);
-  }
-
-  set _contentType(MediaType value) {
-    headers['content-type'] = value.toString();
-  }
-
-  /// Throw an error if this request has been finalized.
-  void _checkFinalized() {
-    if (!finalized) return;
-    throw new StateError("Can't modify a finalized Request.");
+  /// [body] is the request body. It may be either a [String], a [List<int>], a
+  /// [Stream<List<int>>], or `null` to indicate no body.
+  Request change(
+      {Map<String, String> headers,
+      Map<String, Object> context,
+      dynamic body}) {
+    final updatedHeaders = updateMap/*<String, String>*/(this.headers, headers);
+    final updatedContext = updateMap/*<String, Object>*/(this.context, context);
```

Comment:
Are these explicit type annotations necessary? Can't it infer the type from the parameters?